crates/subgraph/src/apy.rs

```diff
@@ -0,0 +1,801 @@
+                    .then_some(
+                        net_vol
+                            .get_signed()
+                            .saturating_mul(one_18().get_signed())
+                            .saturating_div(starting_capital.get_signed())
+                            .saturating_mul(one_18().get_signed())
+                            .checked_div(annual_rate(start, end)),
+                    )
+                    .flatten()
+            });
+
+        // this token vault apy
+        token_vaults_apy.push(TokenVaultAPY {
+            id: vol.id.clone(),
+            token: vol.token.clone(),
+            start_time: start,
+            end_time: end,
+            apy,
+            net_vol: net_vol.unwrap_or(ParseUnits::I256(I256::ZERO)).get_signed(),
+            capital: starting_capital
+                .unwrap_or(ParseUnits::I256(I256::ZERO))
+                .get_signed(),
+        });
+    }
+
+    Ok(token_vaults_apy)
+}
+
+/// Calculates an order's pairs' ratios from their last trades in a given list of trades
+/// Trades must be sorted indesc order by timestamp, this is the case if queried from subgraph
+/// using this lib functionalities
+fn get_pairs_ratio(order_apy: &OrderAPY, trades: &[Trade]) -> HashMap<TokenPair, Option<I256>> {
+    let mut pair_ratio_map: HashMap<TokenPair, Option<I256>> = HashMap::new();
+    for input in &order_apy.inputs_token_vault_apy {
+        for output in &order_apy.outputs_token_vault_apy {
+            let pair_as_key = TokenPair {
+                input: input.token.clone(),
+                output: output.token.clone(),
+            };
+            let reverse_pair_as_key = TokenPair {
+                input: output.token.clone(),
+                output: input.token.clone(),
+            };
+            // if not same io token and ratio map doesnt already include them
+            if input.token != output.token
+                && !(pair_ratio_map.contains_key(&pair_as_key)
+                    || pair_ratio_map.contains_key(&reverse_pair_as_key))
+            {
+                // find this pairs(io or oi) latest tradetrades from list of order's
+                // trades, the calculate the pair ratio (in amount/out amount) and
+                // its reverse from the latest trade that involes these 2 tokens.
+                // this assumes the trades are already in desc order by timestamp which
+                // is the case when used this lib query to get them
+                let ratio = trades
+                    .iter()
+                    .find(|v| {
+                        (v.input_vault_balance_change.vault.token == input.token
+                            && v.output_vault_balance_change.vault.token == output.token)
+                            || (v.output_vault_balance_change.vault.token == input.token
+                                && v.input_vault_balance_change.vault.token == output.token)
+                    })
+                    .and_then(|latest_trade| {
```

Comment:
can all of this boilerplage for decimals and io ratio extractions for trades and tokens be lifted to impls?

Reply:
surely